Fill.py

```diff
@@ -15,8 +17,71 @@ class FillError(RuntimeError):
     pass
 
 
-def _log_fill_progress(name: str, placed: int, total_items: int) -> None:
-    logging.info(f"Current fill step ({name}) at {placed}/{total_items} items placed.")
+class FillLogger():
+    min_size: int = 1000   # minimum number of items we care about reporting (non-tty)
+    min_time: float = 0.25 # minimum time (in seconds) between reports (tty)
+
+    def __init__(self, total_items: int):
+        self.start_time = time.time()
+        self.prev_time = time.time()
+        self.cur_time = time.time()
+        self.step: int = max(round(total_items * 0.1), 1000)
+        self.total_items: int = total_items
+        self.is_debug: bool   = False
+
+        if logging.getLogger().isEnabledFor(logging.DEBUG):
+            self.is_debug = True
+
+    def log_fill_progress(self, name: str, placed: int, final: bool = False) -> None:
+        # never print the small stuff
+        if self.total_items < self.min_size:
+            return
+
+        self.log_nontty(name, placed, final)
+        if not self.is_debug:
+            self.log_tty(name, placed, final)
+
+    def log_nontty(self, name: str, placed: int, final: bool) -> None:
```

Comment:
i didn't recognize what these method names were supposed to represent at a glance so they should probably be a bit more verbose

Reply:
The original intention was logging to a terminal (tty) or logging to a file (non tty), but debug has muddied that.  

I've been debating renaming them already. Right now i'm thinking "log_minimal" and "log_friendly" but suggestions are welcome.